extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitter.java

```diff
@@ -170,33 +189,83 @@ public void emit(final Event event)
             alertLost.incrementAndGet();
           }
         } else if (event instanceof RequestLogEvent) {
-          if (config.getRequestTopic() == null || !requestQueue.offer(objectContainer)) {
+          if (!config.getEventTypes().contains(KafkaEmitterConfig.EventType.REQUESTS) || !requestQueue.offer(objectContainer)) {
             requestLost.incrementAndGet();
           }
+        } else if (event instanceof SegmentMetadataEvent) {
+          if (!config.getEventTypes().contains(KafkaEmitterConfig.EventType.SEGMENT_METADATA) ) {
+            segmentMetadataLost.incrementAndGet();
+          } else {
+            switch (config.getSegmentMetadataTopicFormat()) {
+              case PROTOBUF:
+                resultBytes = convertMetadataEventToProto((SegmentMetadataEvent) event, segmentMetadataLost);
+                objectContainer = new ObjectContainer<>(
+                    resultBytes,
+                    resultBytes.length
+                );
+                break;
+              case JSON:
+                // Do Nothing. We already have the JSON object stored in objectContainer
+                break;
+              default:
+                throw new UnsupportedOperationException("segment.metadata.topic.format has an invalid value " + config.getSegmentMetadataTopicFormat().toString());
+            }
+            if (!segmentMetadataQueue.offer(objectContainer)) {
+              segmentMetadataLost.incrementAndGet();
+            }
+          }
         } else {
           invalidLost.incrementAndGet();
         }
       }
-      catch (JsonProcessingException e) {
+      catch (Exception e) {
         invalidLost.incrementAndGet();
+        log.warn(e, "Exception while serializing event");
       }
     }
   }
 
+  private byte[] convertMetadataEventToProto(SegmentMetadataEvent event, AtomicLong segmentMetadataLost)
+  {
+    try {
+      Timestamp createdTimeTs = Timestamps.fromMillis(event.getCreatedTime().getMillis());
+      Timestamp startTimeTs = Timestamps.fromMillis(event.getStartTime().getMillis());
+      Timestamp endTimeTs = Timestamps.fromMillis(event.getEndTime().getMillis());
+
+      DruidSegmentEvent.Builder druidSegmentEventBuilder = DruidSegmentEvent.newBuilder()
+          .setDataSource(event.getDatasource())
+          .setCreatedTime(createdTimeTs)
+          .setStartTime(startTimeTs)
+          .setEndTime(endTimeTs)
+          .setVersion(event.getVersion())
+          .setIsCompacted(event.isCompacted());
+      if (config.getClusterName() != null) {
+        druidSegmentEventBuilder.setClusterName(config.getClusterName());
+      }
+      DruidSegmentEvent druidSegmentEvent = druidSegmentEventBuilder.build();
+      return druidSegmentEvent.toByteArray();
+    }
+    catch (Exception e) {
+      log.warn(e, "Exception while serializing SegmentMetadataEvent");
+      segmentMetadataLost.incrementAndGet();
```

Comment:
It looks like we would be double-counting lost events if we are incrementing on exception here as well as in the calling function. Is there a reason we are catching it in both places rather than letting the calling function handle all serialization exceptions, similar to how Json serialization exceptions would be handled?

Reply:
right, but the behavior is not consistent across event types for all formats

If a json event fails to serialize (e.g. alerts), we do not increment `alertsLost`, we only throw an exception which gets caught and causes `invalidLost` to get incremented.

for segment metadata,
- if we use proto, and there  is a proto serialization issue, then we would end up incrementing both `segmentMetadataLost` here and `invalidLost` due to the exception being thrown
- if we use json, and there is a json serialization issue, we would not increment `segmentMetadataLost`, but only `invalidLost`, which is consistent with the other events

So the proto case is not consistent with the other ones and we would double-count an event as both lost and invalid, rather than only one or the other as is the case for other events.